### PR TITLE
Locking refactoring

### PR DESCRIPTION
**Pull request by paulgb**
This PR:
- Adds a `--lock` option to the `plane-cli spawn` command, which provides a way to spawn a backend with a lock.
- Adds a `lock` field to assignment messages and processes locks by storing them in the global state at the cluster level.
- Checks if a schedule request includes a lock. If it does, checks if that lock is held. If it is, returns information about the existing backend instead of spawning a new one.
- Changes `ScheduleResponse` to include a `spawned` bool, which is `true` if a new backend was spawned but false if an existing backend was returned due to a lock conflict.
- Adds a test

Currently, if a backend has a bearer token, that bearer token is not returned in the result, meaning that this is not very useful yet in combination with bearer tokens. Resolving that will require storing the bearer token in the cluster state, which I will do in a future PR. I've opened #369, #370, and #371 as follow-up tasks.

### CODE REVIEW

File: controller/src/lib.rs [hunk 1]
```diff
@@ -21,12 +21,58 @@ pub mod run;
...
 pub async fn run_scheduler(nats: TypedNats, state: StateHandle) -> NeverResult {
-    let scheduler = Scheduler::new(state);
+    let scheduler = Scheduler::new(state.clone());
     let mut schedule_request_sub = nats.subscribe(ScheduleRequest::subscribe_subject()).await?;
     tracing::info!("Subscribed to spawn requests.");
 
     while let Some(schedule_request) = schedule_request_sub.next().await {
         tracing::info!(spawn_request=?schedule_request.value, "Got spawn request");
+
+        if let Some(lock) = &schedule_request.value.lock {
```
Comment (pretentious7) [inline controller/src/lib.rs:31]: I think it'd be easier to read if this were factored into a handle_locked_request or something.

File: dev/tests/scheduler.rs [hunk 2]
```diff
@@ -346,3 +388,46 @@ async fn test_update_backend_stats_message() {
         result.0
     );
 }
+
+#[integration_test]
+async fn schedule_request_lock() {
```
Comment (rolyatmax) [inline dev/tests/scheduler.rs:393]: Should we add a test that shows releasing and creating a new lock works as expected?

File: controller/src/lib.rs [hunk 3]
```diff
@@ -21,12 +21,58 @@ pub mod run;
...
+            tracing::info!(lock=%lock, "Request includes lock.");
+            let locked = {
+                let state = state.state();
+                let cluster = state
+                    .cluster(&schedule_request.value.cluster)
+                    .ok_or_else(|| anyhow!("Cluster does not exist."))?;
```
Comment (rolyatmax) [inline controller/src/lib.rs:37]: If the schedule request includes a cluster that does not exist, will this end up returning and stopping the `run_scheduler` loop?

File: core/src/state/world_state.rs [hunk 4]
```diff
@@ -115,6 +125,31 @@ impl ClusterState {
...
+    pub fn locked(&self, lock: &str) -> Result<Option<BackendId>> {
+        let Some(lock_owner) = self.locks.get(lock) else {
+            // The lock does not exist.
+            return Ok(None)
+        };
+
+        let Some(backend) = self.backends.get(lock_owner) else {
+            // The lock exists, but the backend has been purged.
+            // This must be true, because an assignment message is the only way to create a lock.
+            return Ok(None)
```
Comment (rolyatmax) [inline core/src/state/world_state.rs:138]: When can a backend get purged from the `self.backends` list? Is this an unexpected state? Should we log anything here?

File: core/src/state/world_state.rs [hunk 5]
```diff
@@ -115,6 +125,31 @@ impl ClusterState {
...
         self.backends.get(backend)
     }
+
+    pub fn locked(&self, lock: &str) -> Result<Option<BackendId>> {
+        let Some(lock_owner) = self.locks.get(lock) else {
+            // The lock does not exist.
+            return Ok(None)
+        };
+
+        let Some(backend) = self.backends.get(lock_owner) else {
```
[the remark this answers is not in the document]
Comment (rolyatmax) [reply core/src/state/world_state.rs:135]: Would setting the backend's state to `Lost` be the equivalent?

File: controller/src/lib.rs [hunk 6]
```diff
@@ -21,12 +21,58 @@ pub mod run;
...
+    let scheduler = Scheduler::new(state.clone());
     let mut schedule_request_sub = nats.subscribe(ScheduleRequest::subscribe_subject()).await?;
     tracing::info!("Subscribed to spawn requests.");
 
     while let Some(schedule_request) = schedule_request_sub.next().await {
         tracing::info!(spawn_request=?schedule_request.value, "Got spawn request");
+
+        if let Some(lock) = &schedule_request.value.lock {
+            tracing::info!(lock=%lock, "Request includes lock.");
+            let locked = {
```
[the remark this answers is not in the document]
Comment (paulgb) [reply controller/src/lib.rs:33]: Initially the idea is that there is only one controller running at a time (currently this is enforced outside of Plane, e.g. by kubernetes or just by running Plane manually). `run_scheduler` is invoked once in the controller, and only schedules one thing at a time, and is the only place where a lock can be acquired.

File: controller/src/lib.rs [hunk 7]
```diff
@@ -21,12 +21,58 @@ pub mod run;
...
+                let state = state.state();
+                let cluster = state
+                    .cluster(&schedule_request.value.cluster)
+                    .ok_or_else(|| anyhow!("Cluster does not exist."))?;
+                cluster.locked(lock)?
```
[the remark this answers is not in the document]
Comment (paulgb) [reply controller/src/lib.rs]: Good catch. There is not, so I've removed it.

File: controller/src/lib.rs [hunk 8]
```diff
@@ -46,6 +92,7 @@ pub async fn run_scheduler(nats: TypedNats, state: StateHandle) -> NeverResult {
                                 backend: spawn_request.backend_id.clone(),
                                 message: BackendMessageType::Assignment {
                                     drone: drone_id.clone(),
+                                    lock: schedule_request.value.lock.clone(),
```
[the remark this answers is not in the document]
Comment (paulgb) [reply controller/src/lib.rs:95]: Yes. Seems like the best solution here would be to wait for the world state to update before continuing, but I'll give it some thought as that would also cut down on scheduling throughput unnecessarily.

File: controller/src/lib.rs (hunk 3 above)
Comment (paulgb) [reply controller/src/lib.rs:37]: Yes. It's not the end of the world now that the scheduler runs inside a `Supervisor`, but it should really just log it and continue.

File: controller/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rolyatmax) [reply controller/src/lib.rs:33]: Ah gotcha - so this `run_scheduler` loop only runs in one thread? In other words, even with the `awaits` in there, each iteration of this loop must finish (i.e. each message must be fully processed) before the next iteration can start (i.e. before the next message can begin processing)?

[the remark this answers is not in the document]
Comment (paulgb) [reply controller/src/lib.rs:33]: Yes, exactly.

Technically, it runs in one logical thread within the tokio runtime. The tokio runtime is allowed to move it to a different OS thread if it wants to (e.g. if the thread it was running on before an await is now doing something else).

**Pull request merged**